src/models/Location.java

```diff
@@ -6,97 +6,278 @@
+                null,
+                null,
+                isInitial,
+                isUrgent,
+                isUniversal,
+                isInconsistent,
+                productOf,
+                null
         );
+    }
 
-        this.isInitial = locations.stream().allMatch(location -> location.isInitial);
-        this.isUrgent = locations.stream().anyMatch(location -> location.isUrgent);
-        this.isUniversal = locations.stream().allMatch(location -> location.isUniversal);
-        this.isInconsistent = locations.stream().anyMatch(location -> location.isInconsistent);
+    public static Location createUniversalLocation(
+            String name,
+            boolean isInitial,
+            boolean isUrgent,
+            int x,
+            int y
+    ) {
+        return new Location(
+                name,
+                x,
+                y,
+                new TrueGuard(),
+                null,
+                null,
+                isInitial,
+                isUrgent,
+                true,
+                false,
+                new ArrayList<>(),
+                null
+        );
+    }
 
-        CDD invariant = CDD.cddTrue();
-        for (Location location : locations) {
-            invariant = location.getInvariantCDD().conjunction(invariant);
-            this.x += location.x;
-            this.y = location.y;
-        }
+    public static Location createUniversalLocation(String name, int x, int y) {
+        return Location.createUniversalLocation(name, false, false, x, y);
+    }
 
-        this.invariant = invariant.getGuard();
+    public static Location createInconsistentLocation(
+            String name,
+            boolean isInitial,
+            boolean isUrgent,
+            int x,
+            int y
+    ) {
+        return new Location(
+                name,
+                x,
+                y,
+                new FalseGuard(),
+                null,
+                null,
+                isInitial,
+                isUrgent,
+                false,
+                true,
+                new ArrayList<>(),
+                null
+        );
+    }
 
-        // We use the average location coordinates
-        this.x /= locations.size();
-        this.y /= locations.size();
+    public static Location createInconsistentLocation(String name, int x, int y) {
+        return Location.createInconsistentLocation(name, false, false, x, y);
     }
 
-    public Location(State state, List<Clock> clocks) {
-        this(
-                state.getLocation().getName(),
-                state.getInvariants(clocks),
-                state.getLocation().getIsInitial(),
-                state.getLocation().getIsUrgent(),
-                state.getLocation().getIsUniversal(),
-                state.getLocation().getIsInconsistent(),
-                state.getLocation().getX(),
-                state.getLocation().getX()
+    public static Location createSimple(Location location) {
```

Comment:
I have tried to add an explanation, however it is difficult. Simple locations are rarely used and should ultimately never be used, but for now they are required because pruning makes a lot of assumptions and state changes on the child location.